### ISSUE TITLE
Add `-C <DIR>` flag to `cargo` to change directory before invoking

### ISSUE THREAD

**Issue by lilyball** [rust-lang/cargo#10098]
### Problem

Cargo infers the manifest path and reads config based on the current directory. Plugins may potentially locate other config this same way. Cargo commands offer a `--manifest-path` flag to override the manifest, but this doesn't override config.

For example, given the following:

```
> cat bin/foo/.cargo/config.toml
[build]
target = "wasm32-unknown-unknown"
target-dir = "../../target"
> cargo build --manifest-path bin/foo/Cargo.toml
```

The resulting build will use the host platform instead of `wasm32-unknown-unknown` and the build artifacts will end up in `bin/foo/target`. But were I to run `cd bin/foo && cargo build` it would use the correct target and target dir.

### Proposed Solution

Cargo should have a flag `-C <DIR>` (or `--directory <DIR>`) that exists at the top level rather than in the subcommands. This flag would cause `cargo` to change its working directory to the specified one before processing any commands or loading any config. It would be roughly equivalent to `( cd <DIR> && cargo … )`.

Besides fixing the issue of loading cargo config, this is also a simpler interface for end users than `--manifest-path` as it means they don't have to worry about where the specific `Cargo.toml` file is, they can just run `cargo` as though it were in a specific directory.

This flag is inspired by both `make` and `git` which offer the same `-C <DIR>` (`make` has the long `--directory` form, I don't believe `git` has a long form). The desired behavior here should be based on `git` as it also defines how multiple `-C` flags interact (subsequent ones are interpreted relative to the earlier ones). Like `git`, any other flags or parameters that take relative paths should then be interpreted relative to the working directory caused by the `-C` option.

### Notes

Regarding rustup and `rust-toolchain.toml` overrides, I'm inclined to say rustup should honor this flag as well and interpret it prior to locating an override file. The idea here is that saying `cargo -C $path …` should behave indistinguishably from running `cargo` within that path.

**Comment by lilyball** [comment]
This request was inspired by my surprised at `cargo build --manifest-path bin/foo/Cargo.toml` not applying the config from `bin/foo/.cargo/config.toml`. I considered filing that as a bug, but I'm not sure what the actual correct behavior there should be, whereas a `-C` flag has an obvious answer (and is easier to use).

**Comment by thomcc** [comment]
Yeah, I hit this issue too. I think ideally `--manifest-path` made cargo do the right thing here, although I guess that would be a breaking change, and I can see some use cases for it using the current directory.

**Comment by Eh2406** [comment]
I believe that @joshtriplett have brought this up at cargo meetings before, but I cannot find notes at this time.

**Comment by olson-dan** [comment]
Ran into this today as well, let me just echo that it's pretty unfortunate that using `--manifest-path` produces a different and potentially incorrect build depending on where it's invoked from.   I'd prefer if config location started in the directory of the manifest than a new config option.

Potentially this might not be breaking, because anyone depending on the current behavior would have to be building from the root of the project anyway.

**Comment by lilyball** [comment]
At this point, I think that the current behavior of `--manifest-path` is actually correct. The `.cargo/config.toml` file controls the behavior of cargo itself, it's not crate configuration, and so the location I invoke cargo from should be what determines how to find the cargo config. Having a separate `-C <DIR>` flag solves this. Especially the part about toolchain overrides (e.g. `rust-toolchain.yaml`) as that is explicitly based on your current path.

**Comment by compenguy** [comment]
> At this point, I think that the current behavior of `--manifest-path` is actually correct. The `.cargo/config.toml` file controls the behavior of cargo itself, it's not crate configuration, and so the location I invoke cargo from should be what determines how to find the cargo config.

This is contrary to the [cargo config documentation](https://doc.rust-lang.org/cargo/reference/config.html), which says:

> With this structure, you can specify configuration per-package, and even possibly check it into version control. You can also specify personal defaults with a configuration file in your home directory.

the per-package .cargo/config.toml checked into version control is being ignored when built via `--manifest-path`, unless the cwd is currently somewhere within the package directory tree, but the primary use case for `--manifest-path` is to make build independent of cwd.  This is either a cargo bug, or a cargo documentation bug, take your pick, but my vote is for cargo bug.

**Comment by epage** [comment]
btw https://github.com/rust-lang/cargo/issues/2930 is the issue about the config being relative to CWD rather than `--manifest-path`.  This issue is more about acknowledging things as they currently are and providing a mechanism for dealing with it

> the per-package .cargo/config.toml checked into version control is being ignored when built via --manifest-path, unless the cwd is currently somewhere within the package directory tree, but the primary use case for --manifest-path is to make build independent of cwd. This is either a cargo bug, or a cargo documentation bug, take your pick, but my vote is for cargo bug.

I could see there being some minor wording tweaks to the documentation.  My guess is its written assuming auto-discovery of the manifest in which case, it is correct.  That assumption should be made more clear though.

**Comment by weihanglo** [comment]
Related issue:

- https://github.com/rust-lang/cargo/issues/9123

**Comment by ehuss** [comment]
Reopening since in #11960 we are planning to mark -C as unstable for now.

**Comment by joshtriplett** [comment]
Nominating to consider whether we want to have `rust-toolchain.toml` support be a blocker. I don't think there's any reasonable way we could support that.

**Comment by weihanglo** [comment]
t-rustup membership has changed a lot since the feature was added. Wonder if they have any new ideas. 

cc @rami3l

**Comment by rami3l** [comment]
@weihanglo Thanks for the ping. I think on our side nothing has changed for `rust-toolchain.toml` since that feature is added, and I don't see any way to support this unless we do a very dirty pattern matching against `cargo`'s argument list. So personally I don't see how we can push this forward (the user has to explicitly instruct `rustup` instead of adding extra instructions to `cargo`) without leaking abstractions, sorry 🙏

**Comment by joshtriplett** [comment]
cc @rami3l @rust-lang/rustup 

We talked about this in today's @rust-lang/cargo meeting. We came up with ~4 options for handling this and moving forward:

1) Cargo could declare that the `-C` option can *only* appear first on the command line, and then we could teach rustup to handle `-C directory` (or `-Cdirectory` or `-C=directory`) with that position restriction to make it simple. We don't *love* having to restrict the position of an argument, but could potentially live with it.

2) rustup could find the `-C` option anywhere it can currently appear (up to the point where the subcommand occurs). We're not super excited about whether this would be sufficiently robust about future command-line options or unusual command-line option handling (as one of many potential examples: if we ever had an option that accepted a mandatory argument that could start with a dash, such that `-X -C abc` is actually passing `-C` to `-X` and then using the subcommand `abc`). But it's worth considering.

3) If we get and process the `-C` option, and we were run via rustup, we could try to re-invoke the rustup wrapper in the new directory. This has a variety of issues, including performance, and the possibility of running a rust-toolchain file in the wrong directory and *then* running another one in the correct directory.

4) We could decide that `-C` does not need to wory about `rust-toolchain.toml`. This would not require any changes, and would work for most people, but could potentially lead to surprises.

**Comment by ChrisDenton** [comment]
Rustup really really doesn't want to be in the business of parsing other process' arguments if we can at all avoid it. We want the wrappers to be as lightweight as possible and, as you say, there are robustness issues to consider.

I think 4 is the best option of those listed. Maybe rustup could have its own special syntax for setting the current directory if people want that (or maybe an option on `rustup run`) but I don't feel strongly about it.

**Comment by weihanglo** [comment]
@ChrisDenton For option 4, those surprises could lead to more issues being filed against `rust-lang/rustup` repo, even though `-C` is a Cargo flag. We also want to make sure t-rustup is fully aware of and aligned with this option, since it affects rustup's user experience, not just Cargo.

**Comment by ChrisDenton** [comment]
For sure, it's a least worst option rather than a perfect solution. I think documenting this case will help. Maybe cargo could also warn if a `rust-toolchain.toml` is present but I fear that might be too noisy. 

But this is all hypothetical at this point, right? We don't really have a good way to gauge how this will play out if few people are using it atm?

**Comment by epage** [comment]
> Maybe cargo could also warn if a rust-toolchain.toml is present but I fear that might be too noisy.

Cargo generally limits its warnings to when a user has a reasonable way to turn it off. We are adding a linting system but this would be outside of that.

This would also have false positives because the file could lead to the same result.

We also wouldn't know about config overrides.

**Comment by rami3l** [comment]
@weihanglo I still feel strange about this: it feels like either rustup or cargo needs to intercept this flag, but not both, since the behavior of both would rely on the current working directory (specifically, in rustup, this would influence directory overrides via both `rust-toolchain.toml` and `rustup override set`).

Once `cd` has been performed it should influence both, so I think it probably should be a `rustup` feature rather than a `cargo` one, meaning `rustup` would take `-C` alongside `+toolchain` (taking the option 1 in https://github.com/rust-lang/cargo/issues/10098#issuecomment-5186156426) and `cd` before doing anything else.

However, considering the fact that people might be using `cargo` without `rustup`, there might still be a use case for `cargo -C`.

Thus, we may end up with both tools supporting this option, with `cargo -C` being rarely used by the user. In the end I doubt whether that would be the best way forward.

**Comment by djc** [comment]
I'm not a fan of the plan of supporting `-C` and just breaking people who rely on `rust-toolchain.toml`.

I have wanted this feature many times, but I'd rather come up with a better plan to support it well instead of having an implementation that breaks a bunch of expectations.

**Comment by ChrisDenton** [comment]
I think if we want to switch `rust-toolchain.toml` via the command line then we need to properly support this in rustup itself.

**Comment by djc** [comment]
Yup. Maybe we could communicate whether/how we changed the current working dir to Cargo via an environment variable?

**Comment by rami3l** [comment]
> Yup. Maybe we could communicate whether/how we changed the current working dir to Cargo via an environment variable?

@djc That'd make perfect sense to me, but I think the main question would be how to capture the user’s intent in rustup.

**Comment by ChrisDenton** [comment]
In any case we probably should take a moment to explorer the design space on the rustup side and then come back to Cargo.

**Comment by djc** [comment]
> In any case we probably should take a moment to explorer the design space on the rustup side and then come back to Cargo.

I don't see many downsides to doing so in this issue, where everyone can keep track of what's going on?

Here's my straw man proposal:

If rustup in proxy mode sees an invocation of `cargo` with a `-C` option passed in, we change the working dir for the child process and set `RUSTUP_SET_CWD` containing the path that we set in the child's environment.

**Comment by rami3l** [comment]
> If rustup in proxy mode sees an invocation of `cargo` with a `-C` option passed in, we change the working dir for the child process and set `RUSTUP_SET_CWD` containing the path that we set in the child's environment.

@djc I'm not sure about this, considering the fact that rustup fully delegates the resolution of `RUSTUP_SET_CWD` to cargo will cause the inverse case scenario compared the one proposed above (where `cargo` supports `-C` but `rustup` is not informed): now `rustup` will recognize the new CWD but `cargo` may or may not act accordingly, depending on how recent it is.

**Comment by djc** [comment]
> [@djc](https://github.com/djc) I'm not sure about this, considering the fact that rustup fully delegates the resolution of `RUSTUP_SET_CWD` to cargo will cause the inverse case scenario compared the one proposed above (where `cargo` supports `-C` but `rustup` is not informed): now `rustup` will recognize the new CWD but `cargo` may or may not act accordingly, depending on how recent it is.

If we still pass `-C` on to Cargo as well (which I didn't articulate in my previous message), it would make older Cargo fail, while newer Cargo can act accordingly. Seems okay?
